xtask/src/flash_image.rs

```diff
@@ -0,0 +1,504 @@
+// Licensed under the Apache-2.0 license
+
+use crate::DynError;
+use crc32fast::Hasher;
+use std::fs::File;
+use std::io::{self, Read, Write};
+use std::io::{Error, ErrorKind};
+
+const FLASH_IMAGE_MAGIC_NUMBER: u32 = 0x464C5348;
+const HEADER_VERSION: u16 = 0x0001;
+const CALIPTRA_FMC_RT_IDENTIFIER: u32 = 0x00000001;
+const SOC_MANIFEST_IDENTIFIER: u32 = 0x00000002;
+const MCU_RT_IDENTIFIER: u32 = 0x00000002;
+const SOC_IMAGES_BASE_IDENTIFIER: u32 = 0x00001000;
+
+pub struct FlashImage {
+    header: FlashImageHeader,
+    checksum: FlashImageChecksum,
+    payload: FlashImagePayload,
+}
+
+pub struct FlashImageHeader {
+    magic_number: u32,
+    header_version: u16,
+    image_count: u16, // number of images
+}
+
+pub struct FlashImageChecksum {
+    header: u32,  // checksum of the header
+    payload: u32, // checksum of the payload
+}
+
+pub struct FlashImagePayload {
+    image_info: Vec<FlashImageInfo>,
+    images: Vec<FirmwareImage>,
+}
+
+// Per image header
+pub struct FlashImageInfo {
+    identifier: u32,
+    image_offset: u32, // Location of the image in the flash as an offset from the header
+    size: u32,         // Size of the image
+}
+
+#[derive(Clone)]
+pub struct FirmwareImage {
+    identifier: u32,
+    data: Vec<u8>,
+}
+
+impl FirmwareImage {
+    pub fn new(identifier: u32, filename: &str) -> io::Result<Self> {
+        let mut file = File::open(filename)?;
+        let mut data = Vec::new();
+        file.read_to_end(&mut data)?;
+
+        Ok(Self { identifier, data })
+    }
+}
+
+impl FlashImage {
+    pub fn new(images: &mut [FirmwareImage]) -> Self {
+        let mut image_info = Vec::new();
+        let mut offset = std::mem::size_of::<FlashImageHeader>() as u32
+            + std::mem::size_of::<FlashImageChecksum>() as u32
+            + (std::mem::size_of::<FlashImageInfo>() * images.len()) as u32;
+
+        for image in images.iter_mut() {
+            let image_size = image.data.len() as u32;
+            Self::align_to_4_bytes(&mut image.data);
+            let padded_size = image.data.len() as u32;
+            image_info.push(FlashImageInfo {
+                identifier: image.identifier,
+                image_offset: offset,
+                size: image_size,
+            });
+            offset += padded_size;
+        }
+
+        let header = FlashImageHeader {
+            magic_number: FLASH_IMAGE_MAGIC_NUMBER,
+            header_version: HEADER_VERSION,
+            image_count: image_info.len() as u16,
+        };
+
+        let payload = FlashImagePayload {
+            image_info,
+            images: images.to_owned(),
+        };
+
+        let checksum = FlashImageChecksum::new(&header, &payload);
+
+        Self {
+            header,
+            checksum,
+            payload,
+        }
+    }
+
+    fn align_to_4_bytes(data: &mut Vec<u8>) {
+        let padding = (4 - (data.len() % 4)) % 4;
```

Comment:
For readability, I might suggest:

```suggestion
        let padding = data.len().next_multiple_of(4) - data.len();
```